### PR TITLE
plan, ranger: support `not in` and correct the behavior.

### PR DESCRIPTION
**Pull request by winoros**
Support to use expression like `a not in (1, 2, 3)` to calculate range.
And for `a in (NULL, 1, 2, 3)`, we shouldn't consider `[<nil>, <nil>]` as a part of range.
PTAL @hanfei1991 @lamxTyler @coocood @zz-jason

### CODE REVIEW

**Comment by winoros** [comment]
need to consider more carefully.

**Comment by winoros** [comment]
PTAL @coocood @hanfei1991

File: util/ranger/ranger_test.go [hunk 1]
```diff
@@ -391,14 +403,28 @@ func (s *testRangerSuite) TestIndexRange(c *C) {
...
+			accessConds: "[in(test.t.c, 1, 1, 1, 1, 1, 1, 2, 1, 2, 3, 2, 3, 4, 4, 1, 2)]",
+			filterConds: "[]",
+			resultStr:   "[[1,1] [2,2] [3,3] [4,4]]",
+		},
+		{
+			indexPos:    1,
+			exprStr:     "c not in (1, 2, 3)",
+			accessConds: "[not(in(test.t.c, 1, 2, 3))]",
+			filterConds: "[]",
+			resultStr:   "[(<nil> +inf,1 <nil>) (1 +inf,2 <nil>) (2 +inf,3 <nil>) (3 +inf,+inf +inf]]",
```
Comment (birdstorm) [inline util/ranger/ranger_test.go]: how about more tests concerning `not in null`

File: util/ranger/ranger_test.go [hunk 2]
```diff
@@ -259,6 +259,18 @@ func (s *testRangerSuite) TestTableRange(c *C) {
...
+			exprStr:     "a in (1, 1, 1, 1, 1, 1, 2, 1, 2, 3, 2, 3, 4, 4, 1, 2)",
+			accessConds: "[in(test.t.a, 1, 1, 1, 1, 1, 1, 2, 1, 2, 3, 2, 3, 4, 4, 1, 2)]",
+			filterConds: "[]",
+			resultStr:   "[[1,1] [2,2] [3,3] [4,4]]",
+		},
+		{
+			exprStr:     "a not in (1, 2, 3)",
+			accessConds: "[not(in(test.t.a, 1, 2, 3))]",
+			filterConds: "[]",
+			resultStr:   "[[-9223372036854775807,0] [4,+inf)]",
```
Comment (zimulala) [inline util/ranger/ranger_test.go]: Why not `[-inf,0]`?

Comment (birdstorm) [reply util/ranger/ranger_test.go]: and it shouldn't be -9223372036854775807 either way, most possibly a bug?

**Comment by zimulala** [comment]
LGTM

File: util/ranger/points.go [hunk 3]
```diff
@@ -284,19 +284,24 @@ func (r *builder) buildFromIsFalse(expr *expression.ScalarFunction, isNot int) [
 	return []point{startPoint, endPoint}
 }
 
-func (r *builder) newBuildFromIn(expr *expression.ScalarFunction) []point {
-	var rangePoints []point
+func (r *builder) newBuildFromIn(expr *expression.ScalarFunction) ([]point, bool) {
```
Comment (hanfei1991) [inline util/ranger/points.go]: Why is it named `newBuildFromIn` instead of `BuildFromIn` ?

Comment (winoros) [reply util/ranger/points.go]: Maybe there is a time that there're two `buildFromIn` method...

**Pull request merged**